src/twisted/test/test_twistd.py

```diff
@@ -865,6 +899,47 @@ def fakeSetupEnvironment(self, chroot, rundir, nodaemon, umask,
             ['/foo/chroot', '/foo/rundir', True, 56, '/foo/pidfile'])
 
 
+    def test_shedPrivilegesError(self):
+        """
+        Handle errors when calling
+         L{twisted.scripts._twistd_unix.shedPrivileges}
+        """
+        def switchUIDFail(euid, uid, gid):
+            raise OSError(errno.EBADF, "fake")
+
+        def switchUIDPass(euid, uid, gid):
+            """
+            Ignore everything and pass
+            """
+
+        runner = UnixApplicationRunner({})
+        switchUID = self.patch(_twistd_unix, 'switchUID', switchUIDFail)
+        exc = self.assertRaises(SystemExit, runner.shedPrivileges, 35,
+                                200, None)
+        self.assertEqual(exc.code, 1)
+        switchUID.restore()
+        switchUID = self.patch(_twistd_unix, 'switchUID', switchUIDPass)
+        runner.shedPrivileges(35, 200, None)
+        switchUID.restore()
+
+
+    def test_uidWithoutGid(self):
+        """
+        Set the uid without settig the gid and then call
+        L{UnixApplicationRunner.startApplication}.
+        """
+        options = twistd.ServerOptions()
+        options.parseOptions([
+            '--nodaemon',
+            '--uid', os.getuid()])
+        application = service.Application("test_setupEnvironment")
+        self.runner = UnixApplicationRunner(options)
+        runner = UnixApplicationRunner(options)
+        runner.startApplication(application)
+        warningsShown = self.flushWarnings()
+        self.assertEqual(1, len(warningsShown))
```

Comment:
This test should make an assertion about the contents of the warnings, since this is a side effect of `shedPrivileges` that might change.

Reply:
Fixed